utils/cgnetcls/internal/indexallocator/indexallocator.go

```diff
@@ -0,0 +1,62 @@
+// The index allocator is a singleton because this reflects the marks assigned on the filesystem
+
+package indexallocator
+
+import (
+	"errors"
+	"sync"
+)
+
+type allocator struct {
+	size           int
+	startIndex     int
+	availableMarks map[int]struct{}
+	sync.Mutex
+}
+
+var indexes *allocator = &allocator{
+	size:       0,
+	startIndex: 0,
+}
+var once sync.Once
+
+// New create a new indexallocator
+func New(size int, startIndex int) (IndexAllocator, int, int) {
+
+	if indexes.size != 0 {
+		return indexes, indexes.size, indexes.startIndex
+	}
+	once.Do(func() {
+		indexes = &allocator{
+			size:           size,
+			startIndex:     startIndex,
+			availableMarks: make(map[int]struct{}, size),
+		}
+		for i := startIndex; i < startIndex+size; i++ {
+			indexes.availableMarks[i] = struct{}{}
+		}
+	})
+	return indexes, size, startIndex
+}
+
```

Comment:
I think it's much easier to do this with a channel? And we already have an allocator doing the same thing. So I am confused as to why we are doing this again?